internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1033 @@ package controller
+	r.log.Info("Enter Reconcile", "req", client.ObjectKeyFromObject(&r.driver))
+
+	// Load the driver desired state based on driver resource, operator config resource and default values.
+	if err := r.LoadAndValidateDesiredState(); err != nil {
+		return ctrl.Result{}, err
+	}
 
-	// TODO(user): your logic here
+	// Concurrently reconcile different aspects of the clusters actual state to meet
+	// the desired state defined on the driver object
+	errChan := utils.RunConcurrently(
+		r.reconcileK8sCsiDriver,
+		r.reconcileControllerPluginDeployment,
+		r.reconcileNodePluginDeamonSet,
+		r.reconcileLivnessService,
+	)
+
+	// Check if any reconcilatin error where raised during the concurrent execution
+	// of the reconciliation steps.
+	errList := utils.ChannelToSlice(errChan)
+	if err := errors.Join(errList...); err != nil {
+		r.log.Error(err, "Reconciliation failed")
+		return ctrl.Result{}, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *DriverReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&csiv1alpha1.Driver{}).
-		Complete(r)
+func (r *driverReconcile) LoadAndValidateDesiredState() error {
+	// Validate that the requested name for the CSI driver isn't already claimed by an existing CSI driver
+	// (Can happen if a driver with an identical name was created in a different namespace)
+	if err := r.Get(r.ctx, client.ObjectKey{Name: r.driver.Name}, &storagev1.CSIDriver{}); err == nil {
+		r.log.Error(err, "Desired name already in use by a different CSI Driver", "name", r.driver.Name)
+		return nil
+	} else if !k8sErrors.IsNotFound(err) {
+		r.log.Error(err, "Failed to query the existence of a CSI Driver", "name", r.driver.Name)
+		return err
+	}
+
+	// Load operator configuration resource
+	opConfig := csiv1a1.OperatorConfig{}
+	opConfig.Name = operatorConfigName
+	opConfig.Namespace = operatorNamespace
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&opConfig), &opConfig); client.IgnoreNotFound(err) != nil {
+		r.log.Error(err, "Unable to load operatorconfig.csi.ceph.io", "name", client.ObjectKeyFromObject(&opConfig))
+		return err
+	}
+
+	// Extract the driver sort name and driver type
+	matches := nameRegExp.FindStringSubmatch(r.driver.Name)
+	if len(matches) != 3 {
+		return fmt.Errorf("invalid driver name")
+	}
+	r.driverPrefix = matches[1]
+	r.driverType = DriverType(strings.ToLower(matches[2]))
+
+	// Load the current desired state in the form of a ceph csi driver resource
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&r.driver), &r.driver); err != nil {
+		r.log.Error(err, "Unable to load driver.csi.ceph.io object", "name", client.ObjectKeyFromObject(&r.driver))
+		return client.IgnoreNotFound(err)
+	}
+
+	// Creating a copy of the driver spec, making sure any local changes will not effect the object residing
+	// in the client's cache
+	r.driver.Spec = *r.driver.Spec.DeepCopy()
```

Comment:
why a deepcopy here? afaik, `client.Get` fills in the memory that we supply and this is already a copy.

Reply:
The struct contains multiple pointers and Tt the best of my knowledge,`client.Get` is only giving back a shallow copy of the item from the cache (for cached clients)

Reply:
I'll take a look, just a thought, if we get a resource and update it and only the updated object will have new updates however newly get resource will not have new updates as they are fetched from cache again so my understanding whatever we pass will be filled not referred.

Reply:
I'm not able to test it, I had a sample controller and testing setup but it was wiped out recently, failing to recreate it due to update in operator-sdk.

however, I still believe you don't need a deep copy here and did a quick check in code https://github.com/kubernetes-sigs/controller-runtime/blob/1ed345090869/pkg/cache/internal/cache_reader.go#L84-L90

``` go
	if c.disableDeepCopy {
		// skip deep copy which might be unsafe
		// you must DeepCopy any object before mutating it outside
	} else {
		// deep copy to avoid mutating cache
		obj = obj.(runtime.Object).DeepCopyObject()
	}
```

btw I didn't traverse all the code. you could do a simple test as well.